app/scripts/lib/multichain-api/caip25permissions.test.ts

```diff
@@ -215,5 +234,397 @@ describe('endowment:caip25', () => {
+      MockScope.validateAndFlattenScopes.mockReturnValue({
+        flattenedRequiredScopes: 'flattenedRequiredScopes',
+        flattenedOptionalScopes: 'flattenedOptionalScopes',
+      });
+      try {
+        validator({
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: {
+                requiredScopes: {
+                  'eip155:1': {
+                    methods: ['eth_chainId'],
+                    notifications: [],
+                    accounts: ['eip155:1:0xdead'],
+                  },
+                },
+                optionalScopes: {
+                  'eip155:5': {
+                    methods: [],
+                    notifications: [],
+                    accounts: ['eip155:5:0xbeef'],
+                  },
+                },
+                isMultichainOrigin: true,
+              },
+            },
+          ],
+          date: 1234,
+          id: '1',
+          invoker: 'test.com',
+          parentCapability: Caip25EndowmentPermissionName,
+        });
+      } catch (err) {
+        // noop
+      }
+      expect(MockScope.assertScopesSupported).toHaveBeenCalledWith(
+        'flattenedOptionalScopes',
+        expect.objectContaining({
+          isChainIdSupported: expect.any(Function),
+        }),
+      );
+      const isChainIdSupportedBody =
+        MockScope.assertScopesSupported.mock.calls[1][1].isChainIdSupported.toString();
+      expect(isChainIdSupportedBody).toContain('findNetworkClientIdByChainId');
+    });
+
+    it('throws if the input requiredScopes ScopesObject is not already validated and flattened', () => {
+      MockScope.validateAndFlattenScopes.mockReturnValue({
+        flattenedRequiredScopes: {},
+        flattenedOptionalScopes: {
+          'eip155:5': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:5:0xbeef'],
+          },
+        },
+      });
+      expect(() => {
+        validator({
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: {
+                requiredScopes: {
+                  'eip155:1': {
+                    methods: ['eth_chainId'],
+                    notifications: [],
+                    accounts: ['eip155:1:0xdead'],
+                  },
+                },
+                optionalScopes: {
+                  'eip155:5': {
+                    methods: [],
+                    notifications: [],
+                    accounts: ['eip155:5:0xbeef'],
+                  },
+                },
+                isMultichainOrigin: true,
+              },
+            },
+          ],
+          date: 1234,
+          id: '1',
+          invoker: 'test.com',
+          parentCapability: Caip25EndowmentPermissionName,
+        });
+      }).toThrow(/Expected values to be strictly deep-equal/u);
+    });
+
+    it('throws if the input optionalScopes ScopesObject is not already validated and flattened', () => {
```

Comment:
Oh is the idea here that the `validateAndFlattenScopes` mock is returning an empty object for `flattenedOptionalScopes ` because the optionalScopes were invalid?

Reply:
Maybe the wording of these tests could be a bit more clear? "Throws if the input does not match the output of `validateAndFlattenScopes`"?

Reply:
> Oh is the idea here that the validateAndFlattenScopes mock is returning an empty object for flattenedOptionalScopes  because the optionalScopes were invalid?

right. invalid OR just not flattened, but the point is that the input ScopesObject does not match

> Maybe the wording of these tests could be a bit more clear? "Throws if the input does not match the output of validateAndFlattenScopes"?

I thought about this, but it seems too descriptive / describes the wrong thing? I mean it's not wrong, it just seems that our goal is to throw when the input ScopesObject is not validated and flattened. How we achieve that is an implementation detail.

Reply:
Yeah makes sense. I feel like not everyone feels the same way about naming/describing tests. Personally when the test name doesn't clearly/directly describe what's being tested I like to add inline comments to explain the discrepancy. But its not strictly necessary. And not everyone agrees with this thinking.

Reply:
Fair. It's unclear from the test setup why the mock return value of `validateAndFlattenScopes` must be changed. I've updated the scenario descriptions because of that https://github.com/MetaMask/metamask-extension/pull/26698/commits/bcd8ebcb2f434f9cd2565edbdc92c82c3856711e